### PR TITLE
[GlobalOptimization] Add a pass to do horizontal fusion of contraction operations with a common operand.

### PR DESCRIPTION
**Pull request by pashu123**
This pass does a one-off horizontal fusion of contraction operations
with one common operand. It handles the case where the contraction
might be followed by a truncate operation (when the contraction is
done at higher precision, but final result is in lower precision
typical of quantized models).

The pass is off by default and can be enabled by
`--iree-global-opt-fuse-horizontal-contractions`.

Note this pass might be better off being in Preprocessing.

Co-authored-by: Quinn Dawkins <quinn.dawkins@gmail.com>
Co-authored-by: MaheshRavishankar <mahesh@nod-labs.com>

### CODE REVIEW

**Review by MaheshRavishankar — DISMISSED** [review]
Thanks @pashu123  . This pass need better documentation before landing in main. Let us know if you want one of us to do that, or if you have enough understanding of this pass to do this.

File: compiler/src/iree/compiler/GlobalOptimization/Passes.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/Passes.cpp]: Do we want this on by default?

Comment (hanhanW) [reply compiler/src/iree/compiler/GlobalOptimization/Passes.cpp]: My understanding is that this is good for GPU, but I'm not sure if it is applied to other backends. We should at least run some benchmarks?

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/GlobalOptimization/Passes.cpp]: No, we dont want it on by default (at least for now)

**Comment by pashu123** [comment]
> Thanks @pashu123 . This pass need better documentation before landing in main. Let us know if you want one of us to do that, or if you have enough understanding of this pass to do this.

I can go ahead and document it. Please suggest the changes. Thanks.

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: This is not needed because we already scope everything within `mlir::iree_compiler::GlobalOptimization`. Also, I'd avoid using-directives.

> I view using-directives as time-bombs, both for the parties that deal in them and the language system. – Ashley Hedberg with apologies to Warren Buffett
>
> tl;dr
> Using-directives (using namespace foo) are dangerous enough to be banned by the [Google style guide](https://google.github.io/styleguide/cppguide.html#Namespaces). Don’t use them in code that will ever need to be upgraded.
>
> If you wish to shorten a name, you may instead use a namespace alias (namespace baz = ::foo::bar::baz;) or a using-declaration (using ::foo::SomeName), both of which are permitted by the style guide in certain contexts (e.g. in *.cc files).

See https://abseil.io/tips/153 for more details.

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: I think we don't need `truncUser` here. It is only used by `getHorizontalFusionGroupMembers`. Instead, can we get the trunc user in the `getHorizontalFusionGroupMembers` method?

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: This pass does not depend on math dialect directly, so we can remove it.

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: This pass does not depend on math dialect directly, so we can remove it.

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp:87]: Do we plan to add `arith::TruncIOp` to the allow list? If not, maybe we can rename the function to `getTruncFUser`

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: IMO `canBeGrouped` should directly take `linalg::LinalgOp` inputs (not `Operation*`). Then you don't need the casting check in the lambda. And here can be:

```
if (!linalgUser || !canBeGrouped(linalgUser)) {
  continue;
}
```

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: I think it is worths to document what the "candidates" means, so others (like me) don't need to look at all the implementation details.

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp:288]: Please add docs and examples for the method because it is not trivial. Also, inlining comments in the implementation can help readability.

I did not expect tensor ops (like concat, reshapes) generated in the pass. I thought that it could create a `flow.dispatch.*` and put all the operations to the dispatch. I realized that it is not the case after I reviewed the code. So I think doc would be very helpful.

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: style nit: use auto because `....::get` already spells the type.

https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: style nit: add comments for `0`.

```suggestion
  Value newRhs = rewriter.create<tensor::ConcatOp>(loc, /*dim=*/0, rhsVals);
```

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: style nit: `auto`

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: should we check any of these is nullptr or not? They could be defined by other operations right?

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp:288]: Could you add some comments to the below code that help understanding the flow of what's happening?

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: Can this be `cast<RankedTensorType>(...)`? The `x.cast<XXX>()` style is being deprecated.

https://mlir.llvm.org/deprecation/

> When casting attributes or type, use the free functions variants, e.g., dyn_cast<T>(x), isa<T>(x), etc. Use of the cast methods variants (e.g., x.dyn_cast<T>()) should be avoided in new code as we’ll remove these methods in the future.

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: nit: this can be `rewriter.getMultiDimIdentityMap(concatTruncType.getRank())`, which looks cleaner.

[optional] can we declare a variable (e.g., `rank = concatTruncType.getRank()`), and use it here and below? It makes code shorter and cleaner.

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp:392]: One of SmallVector constructor takes ArrayRef as inputs. so this can be simplified to `SmallVector<int64_t> shape(fusedResultType.getShape());`

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp:178]: This should also fail on dynamic shapes for the RHS right now because we can't guarantee that they are the same size for all contractions in the group (unless we want to do some analysis, but I think not worth it for now).

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp:300]: nit: Add a comment explaining why we are introducing an extra dimension here to concatenate along.

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp:463]: nit: Add
```suggestion
  // Note: Currently these patterns are required due to early lowering of tensor.concat. When we choose
  // to move the lowering of tensor.concat later, these patterns should be dropped.
  tensor::populateDecomposeTensorConcatPatterns(patterns);
```

or alternatively add a pass for decomposing concat to the pipeline.

**Review by qedawkins — APPROVED** [review]
Only a few comments, otherwise looks good with Hanhan's comments also. Approving so that I'm not blocking once the earlier comments are addressed.

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: Actually only needs tensor and arith dialects.

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp:288]: Added

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp:392]: yeah true... I find this style easier to read.

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp]: Good point.

File: compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp:463]: I decided to not do this right now, but it might be better to just use `tensor.insert_slices` directly. Then you can avoid all the expand shape stuff and just use rank-reduced insert slices.

**Review by hanhanW — APPROVED** [review]

**Pull request merged**
